File: CS/FilterEvent/Form1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Windows.Forms;
using DevExpress.Data.Filtering;
using DevExpress.Data;

namespace FilterEvent
{
    public partial class Form1 : Form
    {
        public Form1()
        {
            InitializeComponent();
        }

        private void Form1_Load(object sender, EventArgs e) {
            DataSet ds = new DataSet();
            ds.ReadXml("nwind.xml");
            this.GridControl.DataSource = ds;
            this.GridControl.DataMember = "Orders";
            this.GridView.ActiveFilterCriteria = CriteriaOperator.Parse("Freight > 0");
        }

        private void GridView_SubstituteFilter(object sender, SubstituteFilterEventArgs e) {
            e.Filter &= CriteriaOperator.Parse("getyear(OrderDate) = ?", Convert.ToInt32(this.beiShowByYear.EditValue));
        }

        private void beiShowByYear_EditValueChanged(object sender, EventArgs e) {
            CriteriaOperator filter = this.GridView.ActiveFilterCriteria;
            this.GridView.BeginDataUpdate();
            try {
                this.GridView.ActiveFilterCriteria = null;
                this.GridView.ActiveFilterCriteria = filter;
            } finally {
                this.GridView.EndDataUpdate();
            }
        }
    }
}

File: VB/FilterEvent/Form1.vb
Imports System
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Windows.Forms
Imports DevExpress.Data.Filtering
Imports DevExpress.Data

Namespace FilterEvent

    Public Partial Class Form1
        Inherits Form

        Public Sub New()
            InitializeComponent()
        End Sub

        Private Sub Form1_Load(ByVal sender As Object, ByVal e As EventArgs)
            Dim ds As DataSet = New DataSet()
            ds.ReadXml("nwind.xml")
            GridControl.DataSource = ds
            GridControl.DataMember = "Orders"
            GridView.ActiveFilterCriteria = CriteriaOperator.Parse("Freight > 0")
        End Sub

        Private Sub GridView_SubstituteFilter(ByVal sender As Object, ByVal e As SubstituteFilterEventArgs)
            e.Filter = e.Filter And CriteriaOperator.Parse("getyear(OrderDate) = ?", Convert.ToInt32(beiShowByYear.EditValue))
        End Sub

        Private Sub beiShowByYear_EditValueChanged(ByVal sender As Object, ByVal e As EventArgs)
            Dim filter As CriteriaOperator = GridView.ActiveFilterCriteria
            GridView.BeginDataUpdate()
            Try
                GridView.ActiveFilterCriteria = Nothing
                GridView.ActiveFilterCriteria = filter
            Finally
                GridView.EndDataUpdate()
            End Try
        End Sub
    End Class
End Namespace
